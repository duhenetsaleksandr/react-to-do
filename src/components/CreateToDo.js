import React, { useEffect, useRef } from 'react';
import '../styles/form.scss';

export default function CreateToDo(props) {
    const inputValue = props.inputValue;
    const { status } = props.editStateInfo;
    const refInputTodo = useRef(null);

    useEffect(() => {
        refInputTodo.current.focus();
        refInputTodo.current.selectionStart = refInputTodo.current.value.length;
        refInputTodo.current.selectionEnd = refInputTodo.current.value.length;
    });

    return (
        <form className="form-add-todo" action="#" onSubmit={props.handlerSubmitForm}>
            <input
                type="text"
                placeholder="Enter new todo"
                value={inputValue}
                onChange={props.handlerChangeInput}
                ref={refInputTodo}
            />
            <button disabled={!inputValue} type="submit">{status ? 'Edit task' : 'Add task'}</button>
        </form>
    );
}
