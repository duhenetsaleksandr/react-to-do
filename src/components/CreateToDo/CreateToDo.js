import React, { useState } from 'react';
import './form.scss';
import InputField from './InputField';

const CreateToDo = ({ editStateInfo, createTodo, editTodo, clearEditMode }) => {
    const { status, todo } = editStateInfo;
    const [ example, setExample ] = useState(true);

    const handlerSubmitForm = (event) => {
        event.preventDefault();
        const { value } = event.target.firstElementChild;
        if (!status) {
            value && createTodo(value);
            setExample(!example);
        } else {
            value && editTodo(todo.id, value);
        }
        clearEditMode();
    }

    return (
        <form className="form-add-todo" name="create-to-do-form" action="#" onSubmit={handlerSubmitForm}>
            <InputField editTodo={todo} onCreated={example} />
            <button type="submit">{status ? 'Edit task' : 'Add task'}</button>
        </form>
    );
}

export default CreateToDo;
