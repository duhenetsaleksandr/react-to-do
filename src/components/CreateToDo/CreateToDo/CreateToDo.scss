@use 'src/styles/colors';

.form-add-todo {
    margin-bottom: 20px;
    display: flex;

    input {
        flex: 1 0 auto;
        margin-right: 20px;
        height: 50px;
        background: none;
        border: 1px solid colors.$secondary-color;
        border-radius: 5px;
        color: colors.$secondary-color;
        padding: 10px 25px;
        font-size: 16px;

        &::placeholder {
            transition: all 0.2s ease;
        }

        &:focus {
            box-shadow: 0 0 5px 0 colors.$secondary-color;
            &::placeholder{
                opacity: 0;
            }
        }
    }

    button {
        width: 150px;
        border-radius: 5px;
        background-color: colors.$secondary-color;
        color: colors.$primary-color;
        font-size: 16px;
        transition: all 0.2s ease;

        &:hover {
            background-color: colors.$helper-color;
        }

        &:active {
            background-color: transparent;
            border: 1px solid colors.$helper-color;
            color: colors.$helper-color;
        }

        &:disabled {
            background-color: colors.$secondary-color;
            opacity: 0.4;
            cursor: not-allowed;
        }
    }
}
