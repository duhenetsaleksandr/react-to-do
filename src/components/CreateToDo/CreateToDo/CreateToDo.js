import React, { useState } from 'react';
import { FormattedMessage } from 'react-intl';
import './CreateToDo.scss';
import InputField from 'components/CreateToDo/InputField';

const CreateToDo = ({ editStateInfo, createTodo, editTodo, clearEditMode }) => {
    const { status, todo } = editStateInfo;
    const [ example, setExample ] = useState(true);

    const handlerSubmitForm = (event) => {
        event.preventDefault();
        const { value } = event.target.firstElementChild;
        if (!status) {
            value && createTodo(value);
            setExample(!example);
        } else {
            value && editTodo(todo.id, value);
        }
        clearEditMode();
    }

    return (
        <form className="form-add-todo" name="create-to-do-form" onSubmit={handlerSubmitForm}>
            <InputField editTodo={todo} onCreated={example} />
            <button type="submit">
                { status ? <FormattedMessage id="button_edit" /> : <FormattedMessage id="button_add" /> }
            </button>
        </form>
    );
}

export default CreateToDo;
