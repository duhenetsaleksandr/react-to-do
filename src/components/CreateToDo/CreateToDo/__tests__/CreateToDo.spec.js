import React from 'react';
import CreateToDo from '../CreateToDo';
import { act } from "@testing-library/react";

const props = {
    editStateInfo: {
        status: false,
        todo: {},
    },
    createTodo: jest.fn(),
    editTodo: jest.fn(),
    clearEditMode: jest.fn(),
};

describe('should render CreateToDo component', () => {
    describe('with default props', () => {
        let component;
        beforeEach(() => {
            component = mount(<CreateToDo {...props} />);
        });

        it('should contain one form', () => {
            const form = component.find('form[name="create-to-do-form"]');
            expect(form).toHaveLength(1);
        });

        it('should contain component InputField', () => {
            const inputField = component.find('InputField');
            expect(inputField).toHaveLength(1);
        });

        it('should contain submit button with text Add task', () => {
            const button = component.find('button[type="submit"]');
            expect(button).toHaveLength(1);
            expect(button.getElement().props.children).toBe('Add task');
        });

        it('should call clearEditMode props function', () => {
            const form = component.find('form[name="create-to-do-form"]');
            act(() => {
                form.getElement().props.onSubmit({ preventDefault: jest.fn(), target: {firstElementChild: {value: ''}} });
            });
            expect(props.clearEditMode).toHaveBeenCalledTimes(1);
        });

        it('should call createTodo props function', () => {
            const form = component.find('form[name="create-to-do-form"]');
            act(() => {
                form.getElement().props.onSubmit({ preventDefault: jest.fn(), target: {firstElementChild: {value: 'example'}} });
            });
            expect(props.createTodo).toHaveBeenCalledWith('example');
        });
    });

    describe('with custom props', () => {
        it('should render button with text Edit task when edit status true', () => {
            const component = mount(<CreateToDo {...props} editStateInfo={{status: true, todo: {}}} />);
            const button = component.find('button[type="submit"]');
            expect(button).toHaveLength(1);
            expect(button.getElement().props.children).toBe('Edit task');
        });

        it('should call editTodo props function', () => {
            const component = mount(<CreateToDo {...props} editStateInfo={{status: true, todo: {id: 1}}} />);
            const form = component.find('form[name="create-to-do-form"]');
            act(() => {
                form.getElement().props.onSubmit({ preventDefault: jest.fn(), target: {firstElementChild: {value: 'example'}} });
            });
            expect(props.editTodo).toHaveBeenCalledWith(1, 'example');
        });

        it('should render CreateToDo component snapshot', () => {
            const component = shallow(<CreateToDo {...props} />);
            expect(component).toMatchSnapshot();
        });
    });
});
