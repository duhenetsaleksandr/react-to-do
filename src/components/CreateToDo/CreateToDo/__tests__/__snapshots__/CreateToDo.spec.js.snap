// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CreateToDo component with custom props should render CreateToDo component snapshot 1`] = `
<form
  className="form-add-todo"
  name="create-to-do-form"
  onSubmit={[Function]}
>
  <InputField
    editTodo={Object {}}
    onCreated={true}
  />
  <button
    type="submit"
  >
    Add task
  </button>
</form>
`;
